File: DSA/patterns/sliding_window/longest_substring.py
#!/bin/python3

'''
    Problem:
    Given a string, find the longest substring with at most k distinct characters.
'''

def longest_substring_k_distinct(s, k):
    '''
        Solution.
    '''
    char_count = {}
    left = 0
    max_length = 0

    for right in range(len(s)):
        char_count[s[right]] = char_count.get(s[right], 0) + 1

        while len(char_count) > k:
            char_count[s[left]] -= 1
            if char_count[s[left]] == 0:
                del char_count[s[left]]
            left += 1

        max_length = max(max_length, right - left + 1)

    return max_length

# Example Usage
print(longest_substring_k_distinct("araaci", 2))  # Output: 4

File: DSA/patterns/sliding_window/longestSubstring.js
function longestSubstringKDistinct(s, k) {
  let charCount = {};
  let left = 0;
  let maxLength = 0;

  for (let right = 0; right < s.length; right++) {
    charCount[s[right]] = (charCount[s[right]] || 0) + 1;

    while (Object.keys(charCount).length > k) {
      charCount[s[left]]--;
      if (charCount[s[left]] === 0) {
        delete charCount[s[left]];
      }
      left++;
    }

    maxLength = Math.max(maxLength, right - left + 1);
  }

  return maxLength;
}

// Example Usage
console.log(longestSubstringKDistinct("araaci", 2)); // Output: 4
